blaze/compute/sparksql.py

```diff
@@ -1,135 +1,136 @@
-    if not isinstance(b, SparkSQLQuery):
-        b = make_query(b)
-    return compute_up(e, a, b, **kwargs)
+    # sub them in the expression
+    expr = expr._subs(dict(zip(tables, new_leaves)))
 
+    # get sqlalchemy tables, we can't go through compute_down here as that will
+    # recurse back into this function
+    sa_tables = [compute_up(t, data) for t in tables]
 
-@dispatch((UnaryOp, Expr), SparkSQLQuery)
-def compute_up(expr, q, **kwargs):
-    scope = kwargs.pop('scope', dict())
-    scope = dict((t, q.mapping.get(data, data)) for t, data in scope.items())
+    # compute using sqlalchemy
+    scope = dict(zip(new_leaves, sa_tables))
+    query = compute(expr, scope)
 
-    q2 = compute_up(expr, q.query, scope=scope, **kwargs)
-    return SparkSQLQuery(q.context, q2, q.mapping)
+    # interpolate params
+    qs = str(literalquery(query))
+    return data.sql(qs)
 
 
-@dispatch((BinOp, Join, Expr), SparkSQLQuery, SparkSQLQuery)
-def compute_up(expr, a, b, **kwargs):
-    assert a.context == b.context
+# see http://spark.apache.org/docs/latest/sql-programming-guide.html#spark-sql-datatype-reference
+sparksql_to_dshape = {
+    sql.ByteType: datashape.int8,
+    sql.ShortType: datashape.int16,
+    sql.IntegerType: datashape.int32,
+    sql.LongType: datashape.int64,
+    sql.FloatType: datashape.float32,
+    sql.DoubleType: datashape.float64,
+    sql.StringType: datashape.string,
+    sql.BinaryType: datashape.bytes_,
+    sql.BooleanType: datashape.bool_,
+    sql.TimestampType: datashape.datetime_,
+    sql.DateType: datashape.date_,
+    # sql.ArrayType: ?,
+    # sql.MapTYpe: ?,
+    # sql.StructType: ?
+}
 
-    mapping = toolz.merge(a.mapping, b.mapping)
 
-    scope = kwargs.pop('scope', dict())
-    scope = dict((t, mapping.get(data, data)) for t, data in scope.items())
+def schema_to_dshape(schema):
+    dshape = []
+    for field in schema.fields:
+        name = field.name
+        value = sparksql_to_dshape.get(type(field.dataType))
+        if value is None:
+            raise ValueError('No known mapping for SparkSQL type %r to '
+                             'datashape type' % str(field.dataType))
+        dshape.append((name, value))
+    return datashape.Record(dshape)
 
-    c = compute_up(expr, a.query, b.query, scope=scope, **kwargs)
-    return SparkSQLQuery(a.context, c, mapping)
 
+def scala_set_to_set(ctx, x):
+    from py4j.java_gateway import java_import
 
-from .sql import select
-def sql_string(query):
-    return pipe(query, select, literalquery, str)
+    # import scala
+    java_import(ctx._jvm, 'scala')
 
-@dispatch(Expr, SparkSQLQuery)
-def post_compute(expr, query, scope=None):
-    result = query.context.sql(sql_string(query.query))
-    if iscollection(expr.dshape) and isscalar(expr.dshape.measure):
-        result = result.map(lambda x: x[0])
-    return result
+    # grab Scala's set converter and convert to a Python set
+    return set(ctx._jvm.scala.collection.JavaConversions.setAsJavaSet(x))
```

Comment:
ha! only if `into` wants to deal with total opaque `JavaObject`s (this comes from `py4j`). I basically had to manually chain the methods one at a time while simultaneously following the docs for each resulting type to find out how to do this.